cfme/common/provider_views.py

```diff
@@ -250,3 +251,148 @@ def is_displayed(self):
+        TAB_NAME = 'Default'
+        api_port = Input('default_api_port')
+        verify_tls = BootstrapSwitch(id='default_tls_verify')
+        ca_certs = Input('default_tls_ca_certs')
+
+    @View.nested
+    class candu(Tab, BeforeFillMixin):  # NOQA
+        TAB_NAME = 'C & U Database'
+        hostname = Input('metrics_hostname')
+        api_port = Input('metrics_api_port')
+        database_name = Input('metrics_database_name')
+        username = Input('metrics_userid')
+        password = Input('metrics_password')
+        confirm_password = Input('metrics_verify')
+        change_password = Text(locator='.//a[normalize-space(.)="Change stored password"]')
+
+        validate = Button('Validate')
+
+
+class OpenStackInfraEndpointForm(View):
+    @View.nested
+    class default(Tab, DefaultEndpointForm, BeforeFillMixin):  # NOQA
+        TAB_NAME = 'Default'
+        api_port = Input('default_api_port')
+        security_protocol = BootstrapSelect('default_security_protocol')
+
+    @View.nested
+    class events(Tab, BeforeFillMixin):  # NOQA
+        TAB_NAME = 'Events'
+        event_stream = RadioGroup(locator='//div[@id="amqp"]')
+        # below controls which appear only if amqp is chosen
+        hostname = Input('amqp_hostname')
+        api_port = Input('amqp_api_port')
+        security_protocol = BootstrapSelect('amqp_security_protocol')
+        change_password = Text(locator='.//a[normalize-space(.)="Change stored password"]')
+
+        username = Input('amqp_userid')
+        password = Input('amqp_password')
+        confirm_password = Input('amqp_verify')
+
+        validate = Button('Validate')
+
+    @View.nested
+    class rsa_keypair(Tab, BeforeFillMixin):  # NOQA
+        TAB_NAME = 'RSA key pair'
+
+        username = Input('ssh_keypair_userid')
+        private_key = FileInput(locator='.//input[@id="ssh_keypair_password"]')
+        change_key = Text(locator='.//a[normalize-space(.)="Change stored private key"]')
+
+
+class ProviderAddView(BaseLoggedInPage):
+    title = Text('//div[@id="main-content"]//h1')
+    name = Input('name')
+    prov_type = BootstrapSelect(id='emstype')
+    api_version = BootstrapSelect(id='api_version')  # only for OpenStack
+    zone = Input('zone')
+
+    add = Button('Add')
+    cancel = Button('Cancel')
+
+    @View.nested
+    class endpoints(View):  # NOQA
+        # this is switchable view that gets replaced with concrete view.
+        # it gets changed according to currently chosen provider type every time
+        # when it is accessed
+        pass
+
+    def __getattribute__(self, item):
+        if item == 'endpoints':
+            # this control is BootstrapSelect in Add view and Text in Edit view
+            provider_type = self.prov_type.selected_option \
+                if isinstance(self.prov_type, BootstrapSelect) else self.prov_type.text
+
+            if provider_type == 'Microsoft System Center VMM':
```

Comment:
One drawback of this change is we loose dependancy between main view and endpoints views. So, If I want to fill forms manually in order to create provider, it will be more sophisticated.
But the benefits are that I managed to get rid of that horrible getattribute function, providers are linked with their endpoints and endpoint forms, etc.